core/src/cas/eav.rs

```diff
@@ -0,0 +1,173 @@
+use error::HolochainError;
+use cas::content::Address;
+use std::collections::HashSet;
+
+/// EAV (entity-attribute-value) data
+/// ostensibly for metadata about entries in the DHT
+/// defines relationships between AddressableContent values
+/// implemented on top of cas::storage::ContentAddressableStorage
+/// @see https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model
+/// Address of AddressableContent representing the EAV entity
+type Entity = Address;
+
+/// using String for EAV attributes (not e.g. an enum) keeps it simple and open
+type Attribute = String;
+
+/// Address of AddressableContent representing the EAV value
+type Value = Address;
+
+// @TODO do we need this?
+// unique (local to the source) monotonically increasing number that can be used for crdt/ordering
+// @see https://papers.radixdlt.com/tempo/#logical-clocks
+// type Index ...
+
+// @TODO do we need this?
+// source agent asserting the meta
+// type Source ...
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct EntityAttributeValue {
+    entity: Entity,
+    attribute: Attribute,
+    value: Value,
+    // index: Index,
+    // source: Source,
+}
+
+impl EntityAttributeValue {
+    pub fn new(entity: &Entity, attribute: &Attribute, value: &Value) -> EntityAttributeValue {
+        EntityAttributeValue{
+            entity: entity.clone(),
+            attribute: attribute.clone(),
+            value: value.clone(),
+        }
+    }
+
+    pub fn entity(&self) -> Entity {
+        self.entity.clone()
+    }
+
+    pub fn attribute(&self) -> Attribute {
+        self.attribute.clone()
+    }
+
+    pub fn value(&self) -> Value {
+        self.value.clone()
+    }
+}
+
+/// eav storage
+/// does NOT provide storage for AddressableContent
+/// use cas::storage::ContentAddressableStorage to store AddressableContent
+/// provides a simple and flexible interface to define relationships between AddressableContent
+pub trait EntityAttributeValueStorage {
+    /// adds the given EntityAttributeValue to the EntityAttributeValueStorage
+    /// append only storage
+    /// eavs are retrieved through constraint based lookups
+    /// @see fetch_eav
+    fn add_eav(&mut self, eav: &EntityAttributeValue) -> Result<(), HolochainError>;
+    /// fetches the set of EntityAttributeValues that match constraints
+    /// None = no constraint
+    /// Some(Entity) = requires the given entity (e.g. all a/v pairs for the entity)
+    /// Some(Attribute) = requires the given attribute (e.g. all links)
+    /// Some(Value) = requires the given value (e.g. all entities referencing an Address)
+    fn fetch_eav(&self, entity: Option<Entity>, attribute: Option<Attribute>, value: Option<Value>) -> Result<HashSet<EntityAttributeValue>, HolochainError>;
+}
```

Comment:
You asked me before, @thedavidmeister, if we would need iterators. Yes we do. @neonphog as added callbacks to our p2p interface draft that enable the network to query complete lists of entries and metas we hold. See https://hackmd.io/_OIEfnTOSzWtW_7sUw6lrg?view.